File: 2096-step-by-step-directions-from-a-binary-tree-node-to-another/2096-step-by-step-directions-from-a-binary-tree-node-to-another.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<char> * path1;
    vector<char> * path2;

    bool dfs(TreeNode* root, vector<char>* path, int value)
    {
        if (root->val==value)
        {
            return true;
        }

        if (root->left!=NULL)
        {
            path->push_back('L');
            if (dfs(root->left, path, value))
            {
                return true;
            }
            path->pop_back();
        }
        if (root->right!=NULL)
        {
            path->push_back('R');
            if (dfs(root->right, path, value))
            {
                return true;
            }
            path->pop_back();
        }

        return false;
    }

    string getDirections(TreeNode* root, int startValue, int destValue) {
        path1 = new vector<char> {};
        path2 = new vector<char> {};

        dfs(root, path1, startValue);
        dfs(root, path2, destValue);

        vector<char> new_path1 = *path1;
        vector<char> new_path2 = *path2;

        int i=0;
        while (i<new_path1.size() && i<new_path2.size())
        {
            if (new_path1[i]==new_path2[i])
            {
                i++;
            }
            else
            {
                break;
            }
        }
        
        string ans="";
        for (int tmp=i; tmp<path1->size(); tmp++)
        {
            ans += "U";
        }
        for (int tmp=i; tmp<new_path2.size(); tmp++)
        {
            ans+=new_path2[tmp];
        }

        return ans;
    }
};

File: 2096-step-by-step-directions-from-a-binary-tree-node-to-another/2096-step-by-step-directions-from-a-binary-tree-node-to-another.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getDirections(self, root: Optional[TreeNode], startValue: int, destValue: int) -> str:
        # Find the path to the node
        def dfs(root, path, value):
            if not root:
                return False

            if root.val==value:
                return True

            if root.left:
                path.append("L")
                if dfs(root.left, path, value):
                    return True
                path.pop()
            if root.right:
                path.append("R")
                if dfs(root.right, path, value):
                    return True
                path.pop()
            
            return False

        path1 = []
        path2 = []

        dfs(root, path1, startValue)
        dfs(root, path2, destValue)

        i = max(len(path1), len(path2))
        
        while len(path1)>0 and len(path2)>0:
            if path1[0]==path2[0]:
                path1.pop(0)
                path2.pop(0)
            else:
                break

        ans = "U"*len(path1)
        for x in path2:
            ans+=x
        return ans